Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -1,20 +1,239 @@
+            SubscribeLocalEvent<ChemMasterComponent, BoundUIOpenedEvent>((_, comp, _) => UpdateUiState(comp));
+
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterSetModeMessage>(OnSetModeMessage);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterSetPillTypeMessage>(OnSetPillTypeMessage);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterReagentAmountButtonMessage>(OnReagentButtonMessage);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterCreatePillsMessage>(OnCreatePillsMessage);
+            SubscribeLocalEvent<ChemMasterComponent, ChemMasterCreateBottlesMessage>(OnCreateBottlesMessage);
+        }
+
+        private void UpdateUiState(ChemMasterComponent chemMaster, bool updateLabel = false)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+            var containerSlot = CompOrNull<ItemSlotsComponent>(chemMaster.Owner)?.Slots.GetValueOrDefault("container_slot");
+            EntityUid? container = containerSlot?.Item;
+            Solution? containerSolution = null;
+
+            if (container.HasValue && container.Value.Valid)
+            {
+                TryComp(container, out FitsInDispenserComponent? fits);
+                _solutionContainerSystem.TryGetSolution(container.Value, fits!.Solution, out containerSolution);
+            }
+
+            var containerName = CompOrNull<MetaDataComponent>(container)?.EntityName;
+            var dispenserName = CompOrNull<MetaDataComponent>(chemMaster.Owner)?.EntityName ?? "";
+            var bufferReagents = bufferSolution.Contents;
+            var bufferCurrentVolume = bufferSolution.CurrentVolume;
+
+            var state = new ChemMasterBoundUserInterfaceState(
+                containerSolution?.CurrentVolume, containerSolution?.MaxVolume, containerName, dispenserName,
+                containerSolution?.Contents, bufferReagents, chemMaster.Mode, bufferCurrentVolume, chemMaster.PillType,
+                chemMaster.PillProductionLimit, chemMaster.BottleProductionLimit, updateLabel
+            );
+            _userInterfaceSystem.TrySetUiState(chemMaster.Owner, ChemMasterUiKey.Key, state);
+        }
+
+        private void OnSetModeMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterSetModeMessage message)
+        {
+            // Ensure the mode is valid, either Transfer or Discard.
+            if (!Enum.IsDefined(typeof(ChemMasterMode), message.ChemMasterMode))
+                return;
+
+            chemMaster.Mode = message.ChemMasterMode;
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void OnSetPillTypeMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterSetPillTypeMessage message)
+        {
+            // Ensure valid pill type. There are 20 pills selectable, 0-19.
+            if (message.PillType > 19)
```

Comment:
this should really be static somewhere. also this doesn't clamp if its <0?

Reply:
It's an `uint` so it can't be <0.